central/sensor/service/pipeline/imageintegrations/pipeline.go

```diff
@@ -226,4 +234,64 @@ func (s *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.M
 	return nil
 }
 
+func (s *pipelineImpl) runRemove(ctx context.Context, id string) error {
+	if err := s.integrationManager.Remove(id); err != nil {
+		return errors.Wrap(err, "removing integration from manager")
+	}
+	return s.datastore.RemoveImageIntegration(ctx, id)
+}
+
+// Run runs the pipeline template on the input and returns the output.
+// Action is currently always update.
+func (s *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.MsgFromSensor, _ common.MessageInjector) error {
+	// Ignore autogenerated registries if they are disabled
+	if autogeneratedRegistriesDisabled {
+		return nil
+	}
+	defer countMetrics.IncrementResourceProcessedCounter(pipeline.ActionToOperation(msg.GetEvent().GetAction()), metrics.ImageIntegration)
+
+	if msg.GetEvent().GetAction() == central.ResourceAction_REMOVE_RESOURCE {
+		return s.runRemove(ctx, msg.GetEvent().GetImageIntegration().GetId())
+	}
+
+	// Extract the cluster name.
+	clusterName, exists, err := s.clusterDatastore.GetClusterName(ctx, clusterID)
+	if err != nil {
+		return errors.Wrapf(err, "error getting cluster name for cluster ID: %s", clusterID)
+	}
+	if !exists {
+		return errox.NotFound.Newf("cluster with id %q does not exist", clusterID)
+	}
+
+	// Set up the integration and update its parameters.
+	imageIntegration := msg.GetEvent().GetImageIntegration()
+	imageIntegration.ClusterId = clusterID
+
+	description, matches, err := setUpIntegrationParams(ctx, imageIntegration)
+	if err != nil {
+		return errors.Wrapf(err, "setting up integration params for %q", imageIntegration.GetId())
+	}
+	source := imageIntegration.GetSource()
+	if source == nil {
+		return s.legacyRun(ctx, imageIntegration, matches)
+	}
+	imageIntegration.Name = fmt.Sprintf("Autogenerated %s for cluster %s from %s/%s", description, clusterName, source.GetNamespace(), source.GetName())
+
+	_, exists, err = s.datastore.GetImageIntegration(ctx, imageIntegration.GetId())
+	if err != nil {
+		return errors.Wrapf(err, "retrieving image integration %q", imageIntegration.GetId())
+	}
+
+	if _, err := s.datastore.AddImageIntegration(ctx, imageIntegration); err != nil {
+		return errors.Wrapf(err, "adding image integration %q", imageIntegration.GetId())
+	}
+	if err := s.integrationManager.Upsert(imageIntegration); err != nil {
+		return errors.Wrapf(err, "notifying of update for image integration %q", imageIntegration.GetId())
+	}
+	if !exists {
```

Comment:
This doesn't make too much sense imho. A modified integration might just as well be a reason to run detection again. It would be nice if we could somehow scope the re-evaluation to images potentially using that integration, but that's far, far beyond the scope of this PR.

I wonder if the concern is that doing this on every update might be too frequent. I don't think however that updates of image pull secrets are very different in frequency from creations - most are probably never updated. A concern is the initial cluster connection. when we receive many new integrations and would start the detection loop several times. To avoid this, it might make sense to add a `hasSynced concurrency.Flag`, that is set to `true` on `Reconcile` followed by triggering detection, and then trigger detection here on every update whenever `s.hasSynced.Get()` is `true`.

Reply:
Hm, not sure I completely get it from your comment, sorry.
Your suggestion is to set the flag within `Reconcile()` and then only short-circuit if the flag is set, right?

Regarding the frequencies, I saw [this comment indicating a "higher" volume of updates due to service accounts](https://github.com/stackrox/stackrox/blob/d14d6e7b1a9a0e4b7c5b0e111cc0c6d330462ffb/central/sensor/service/pipeline/imageintegrations/pipeline.go#L210-L214), but to be fair I'm not quite sure whether this is of any concern.

Reply:
Let me try to rephrase:
- Ideally, if an image integration is updated, we would want to reprocess all deployments with previously unscannable/unretrievable images that potentially use this registry integration.
- As an approximation of the above, we could reprocess all deployments that potentially use this registry integration, unscannable images or not.
- As an approximation of the above, we could reprocess all deployments that are somewhat related to this registry integration (e.g., same cluster), even if they don't use it, and regardless of whether they have unscannable images.

The point is that an updated registry integration can only affect a deployment evaluation if we can now scan an image used by the deployment, that we previously could not scan (due to the content-addressing nature of IDs, "being able to scan an image" is a binary property that we can assume never reverts to `false` once it became `true). Otherwise, there is no information that can be gained from the updated image integration and that could change anything about the evaluation.

The existing code however only reprocesses a deployment when a new image integration is created. This is neither an over- nor under-approximation of the "ideal" handling stated at the top of this post. It **unnecessarily** reprocesses all deployments (even those not using the updated image integration nor using an image from one of the registries this integration covers, it might even reprocess deployments in other clusters, and it also reprocesses deployment for which we already have comprehensive scan results for all images). At the same time, it **misses** cases when a previously existing, erroneous image integration is fixed, and there actually _are_ deployments with still unscanned images due to the previously erroneous image integration. Those will only be fixed after ~1h, whereas those with a _missing_ image integration are fixed directly on creation.

My idea was therefore to just run reprocessing on every image integration update, in order to arrive at something that at least is a consistent over-approximation. Still reprocessing far too much, but at least not missing anything; a fully-overblown but sufficient solution rather than a half-overblown yet still insufficient one.

My suspicion was that the case this approach wasn't chosen was due to the fact that during reconciliation (i.e., a sensor reconnects), we would receive a flood of image integration updates, causing an avalanche of reprocessings. However: (a) this is not how `ShortCircuit()` works anyway, it just triggers a signal, but if the enricher/detector is still busy, a subseqent call will just do nothing (note: this alone makes my comment on setting a flag and calling `ShortCircuit` once after reconciliation pointless, so just ignore it), and (b) the comment you pointed out suggests that this is due to a high volume of (spurious?) updates on OpenShift rather than the reconciliation phase.

So, in summary, keep the existing behavior (which I think now has accidentally been flipped), until we either can do reprocessing in a more targeted way (unscanned images only?), or understand why image integration updates on OpenShift are so noisy.